wire/internal/wire/wire.go

```diff
@@ -661,22 +657,50 @@ func zeroValue(t types.Type, qf types.Qualifier) string {
 	}
 }
 
-// typeVariableName invents a variable name derived from the type name
-// or returns the empty string if one could not be found. There are no
-// guarantees about whether the name is exported or unexported: call
-// export() or unexport() to convert.
-func typeVariableName(t types.Type) string {
+// typeVariableName invents disambiguated variable names derived from the
+// type name.
+// If a name can't be derived from the type, defaultName is used.
+// transform is used to transform the name after it is disambiguated; commonly
+//   used functions are export or unexport.
+// collides is used to see if a name is ambiguous.
+func typeVariableName(t types.Type, defaultName string, transform func(string) string, collides func(string) bool) string {
 	if p, ok := t.(*types.Pointer); ok {
 		t = p.Elem()
 	}
+	var names []string
 	switch t := t.(type) {
 	case *types.Basic:
-		return t.Name()
+		if t.Name() != "" {
+			names = append(names, t.Name())
+		}
 	case *types.Named:
-		// TODO(light): Include package name when appropriate.
-		return t.Obj().Name()
+		obj := t.Obj()
+		if name := obj.Name(); name != "" {
+			names = append(names, name)
+		}
+		// Provide an alternate name prefixed with the package name if possible.
+		// E.g., in case of collisions, we'll use "fooCfg" instead of "cfg2".
+		if pkg := obj.Pkg(); pkg != nil && pkg.Name() != "" {
+			names = append(names, fmt.Sprintf("%s%s", pkg.Name(), strings.Title(obj.Name())))
+		}
+	}
+	if len(names) == 0 {
+		names = append(names, defaultName)
 	}
-	return ""
+
+	// Transform the names.
+	for i, name := range names {
+		names[i] = transform(name)
```

Comment:
It is surprising to me that the `defaultName` is transformed. I would have expected no processing, just disambiguation.